From fbb0700ea7bd9b055ae9e43746a7751be88a8955 Mon Sep 17 00:00:00 2001
From: Konstantin Ilichev <konstantin.ilichev@intel.com>
Date: Wed, 21 Aug 2024 14:55:08 +0000
Subject: [PATCH] Enable FFmpeg support in MCM

---
 configure                | 9 +++++++++
 libavdevice/Makefile     | 5 +++++
 libavdevice/alldevices.c | 5 +++++
 3 files changed, 19 insertions(+)

diff --git a/configure b/configure
index 5af693c954..60caa4ebdf 100755
--- a/configure
+++ b/configure
@@ -285,6 +285,7 @@ External library support:
   --enable-libvo-amrwbenc  enable AMR-WB encoding via libvo-amrwbenc [no]
   --enable-libvorbis       enable Vorbis en/decoding via libvorbis,
                            native implementation exists [no]
+  --enable-mcm             enable Media Communications Mesh library [no]
   --enable-libvpx          enable VP8 and VP9 de/encoding via libvpx [no]
   --enable-libwebp         enable WebP encoding via libwebp [no]
   --enable-libx264         enable H.264 encoding via x264 [no]
@@ -1841,6 +1842,7 @@ EXTERNAL_LIBRARY_LIST="
     jni
     ladspa
     lcms2
+    mcm
     libaom
     libaribcaption
     libass
@@ -3567,6 +3569,11 @@ xwma_demuxer_select="riffdec"
 android_camera_indev_deps="android camera2ndk mediandk pthreads"
 alsa_indev_deps="alsa"
 alsa_outdev_deps="alsa"
+mcm_indev_deps="libmcm_dp"
+mcm_outdev_deps="libmcm_dp"
+mcm_audio_indev_deps="libmcm_dp"
+mcm_audio_outdev_deps="libmcm_dp"
+mcm_audio_pcm16_outdev_deps="libmcm_dp"
 avfoundation_indev_deps="avfoundation corevideo coremedia pthreads"
 avfoundation_indev_suggest="coregraphics applicationservices"
 avfoundation_indev_extralibs="-framework Foundation"
@@ -6692,6 +6699,8 @@ enabled libaribb24        && { check_pkg_config libaribb24 "aribb24 > 1.0.3" "ar
 enabled libaribcaption    && require_pkg_config libaribcaption "libaribcaption >= 1.1.1" "aribcaption/aribcaption.h" aribcc_context_alloc
 enabled lv2               && require_pkg_config lv2 lilv-0 "lilv/lilv.h" lilv_world_new
 enabled libiec61883       && require libiec61883 libiec61883/iec61883.h iec61883_cmp_connect -lraw1394 -lavc1394 -lrom1394 -liec61883
+enabled mcm               && { check_pkg_config libmcm_dp "libmcm_dp >= 24.00" mcm_dp.h mcm_dequeue_buffer ||
+                               die "ERROR: libmcm_dp must be installed and version must be >= 24.00"; }
 enabled libass            && require_pkg_config libass "libass >= 0.11.0" ass/ass.h ass_library_init
 enabled libbluray         && require_pkg_config libbluray libbluray libbluray/bluray.h bd_open
 enabled libbs2b           && require_pkg_config libbs2b libbs2b bs2b.h bs2b_open
diff --git a/libavdevice/Makefile b/libavdevice/Makefile
index c30449201d..62bda757c3 100644
--- a/libavdevice/Makefile
+++ b/libavdevice/Makefile
@@ -20,6 +20,11 @@ OBJS-$(CONFIG_AUDIOTOOLBOX_OUTDEV)       += audiotoolbox.o
 OBJS-$(CONFIG_AVFOUNDATION_INDEV)        += avfoundation.o
 OBJS-$(CONFIG_BKTR_INDEV)                += bktr.o
 OBJS-$(CONFIG_CACA_OUTDEV)               += caca.o
+OBJS-$(CONFIG_MCM_INDEV)                 += mcm_video_rx.o mcm_common.o
+OBJS-$(CONFIG_MCM_OUTDEV)                += mcm_video_tx.o mcm_common.o
+OBJS-$(CONFIG_MCM_AUDIO_INDEV)           += mcm_audio_rx.o mcm_common.o
+OBJS-$(CONFIG_MCM_AUDIO_OUTDEV)          += mcm_audio_tx.o mcm_common.o
+OBJS-$(CONFIG_MCM_AUDIO_PCM16_OUTDEV)    += mcm_audio_tx.o mcm_common.o
 OBJS-$(CONFIG_DECKLINK_OUTDEV)           += decklink_enc.o decklink_enc_c.o decklink_common.o
 OBJS-$(CONFIG_DECKLINK_INDEV)            += decklink_dec.o decklink_dec_c.o decklink_common.o
 OBJS-$(CONFIG_DSHOW_INDEV)               += dshow_crossbar.o dshow.o dshow_enummediatypes.o \
diff --git a/libavdevice/alldevices.c b/libavdevice/alldevices.c
index 8a90fcb5d7..fe0eabdd67 100644
--- a/libavdevice/alldevices.c
+++ b/libavdevice/alldevices.c
@@ -25,6 +25,11 @@
 /* devices */
 extern const AVInputFormat  ff_alsa_demuxer;
 extern const FFOutputFormat ff_alsa_muxer;
+extern const AVInputFormat  ff_mcm_demuxer;
+extern const FFOutputFormat ff_mcm_muxer;
+extern const AVInputFormat  ff_mcm_audio_demuxer;
+extern const FFOutputFormat ff_mcm_audio_muxer;
+extern const FFOutputFormat ff_mcm_audio_pcm16_muxer;
 extern const AVInputFormat  ff_android_camera_demuxer;
 extern const FFOutputFormat ff_audiotoolbox_muxer;
 extern const AVInputFormat  ff_avfoundation_demuxer;
-- 
2.34.1

